compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5109,16 +5109,22 @@ public void visit(BLangElvisExpr elvisExpr, AnalyzerData data) {
         BType actualType = symTable.semanticError;
         if (lhsType != symTable.semanticError) {
             BType referencedType = Types.getReferredType(lhsType);
-            if (referencedType.tag == TypeTags.UNION && referencedType.isNullable()) {
-                BUnionType unionType = (BUnionType) referencedType;
-                LinkedHashSet<BType> memberTypes = unionType.getMemberTypes().stream()
-                        .filter(type -> type.tag != TypeTags.NIL)
-                        .collect(Collectors.toCollection(LinkedHashSet::new));
-
-                if (memberTypes.size() == 1) {
-                    actualType = memberTypes.toArray(new BType[0])[0];
+            if (referencedType.isNullable()) {
+                if (referencedType.tag == TypeTags.UNION) {
+                    BUnionType unionType = (BUnionType) referencedType;
+                    LinkedHashSet<BType> memberTypes = unionType.getMemberTypes().stream()
+                            .filter(type -> type.tag != TypeTags.NIL)
+                            .collect(Collectors.toCollection(LinkedHashSet::new));
+
+                    if (memberTypes.size() == 1) {
+                        actualType = memberTypes.toArray(new BType[0])[0];
+                    } else {
+                        actualType = BUnionType.create(null, memberTypes);
+                    }
                 } else {
-                    actualType = BUnionType.create(null, memberTypes);
+                    // This is a builtin union type (any, anydata, json), we use the type as is
+                    // since we don't have a way to represent the union - nil at the moment.
+                    actualType = referencedType;
```

Comment:
By setting the actual type as reference type, we are unable to perform this kind of assignment. Is there an issue for this?
```ballerina
type Foo boolean|int|float|decimal|string|json[]|map<json>;
function getJson(json i) {
    Foo k = i ?: 10;
}
```

Reply:
This is what I meant by the comment.

Doesn't work similarly with an is check

```ballerina
type Foo boolean|int|float|decimal|string|json[]|map<json>;

function getJson(json i) {
    Foo f;
    if i !is () {
        f = i;
    } else {
        f = 10;
    }
}
```

I don't think there's an issue specifically for this, but I believe we have issues related to narrowing.

Reply:
Was actually able to fix this for `json` and `anydata`, since they are modeled as unions. Updated the PR.